cli/src/commands/run.rs

```diff
@@ -55,16 +53,16 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
     let workspace_command = command.workspace_helper(ui)?;
     let _resolved_commits =
         resolve_multiple_nonempty_revsets(&args.revisions, &workspace_command, ui)?;
-    let _jobs = if let Some(_jobs) = args.jobs {
-        _jobs
-    } else {
-        // Use all available cores
-
-        // SAFETY:
-        // We use a internal constant of 4 threads, if it fails
-        let available =
-            std::thread::available_parallelism().unwrap_or(NonZeroUsize::new(4).unwrap());
-        available.into()
-    };
+    let _jobs = args
+        .jobs
+        .filter(|j| *j > 0)
+        .or_else(|| {
+            // Use all available cores
+            std::thread::available_parallelism()
+                .ok()
+                .and_then(|t| t.try_into().ok())
+        })
+        // Fallback to two cores as we don't know on which system we're running on.
+        .unwrap_or(2usize);
```

Comment:
I still prefer to default to 1 job, regardless of number of cores.

Reply:
- I'm a fan of 1 job by default.
- How are you actually going to use two threads for `jj run`? Do you need to spawn a new workspace to run the extra job? (Maybe we should have left the `--jobs` argument out until we could actually use it.)

Reply:
Thanks for making me look up the implementations for both `tokio::process` and `tokio::fs`. In both platform implementations they suggest that they can't really a process asynchronously (e.g `epoll` / `iocp`), so they listen on best effort for signals. And for creating directories, which is needed for a certain scale of commits, they will be spawned on a background job anyway. 

Initially I partially agreed with one job, but after looking into it, I'd rather do it myself, so we have atleast one controlling thread free.

Reply:
Does the flag control how many concurrent subprocesses we allow or the number of threads to use? I think we should give the user control over the former, but hopefully we don't need to give them control over the latter.

Reply:
> Does the flag control how many concurrent subprocesses we allow or the number of threads to use? I think we should give the user control over the former, but hopefully we don't need to give them control over the latter.

I'd say the latter, as a subprocess w/o a existing directory to work in is useless. Why do you think that controlling processes will help the user? I imagine a user with a threadripper or a beefy 128 core machine will want it done in short a time.

Reply:
See https://github.com/martinvonz/jj/pull/2486#discussion_r1382266739. I didn't understand your reply there.

Reply:
External handholding was always out of scope until now.

Reply:
> * How are you actually going to use two threads for `jj run`? Do you need to spawn a new workspace to run the extra job? 

I'd like to have a thread to control the creation of commit trees.

> (Maybe we should have left the `--jobs` argument out until we could actually use it.)

I'd rather not wait longer on a implementation, as this is incremental in-tree development. We can still change _it_ before users can actually use the command. 

The rest of my replies are inline as edits.

Reply:
I'm fine to proceed with `--jobs` as-is, since we don't use it for anything yet, anyways, but I think there's a fundamental mismatch in meaning about "parallelism":

- One notion is the maximum number of concurrent instances of the user-provided command to run. This corresponds to `git-branchless test`'s `--jobs` argument.
- One notion is the size of some thread pool used to service those jobs, as well as other maintenance overhead, such as checking out a workspace on disk. I don't think the user typically *wants* to configure that value themselves, and, when so, they would probably want to configure it using a more general mechanism that lets you set the size of jj thread pools for various commands.

The design document does not explicitly specify what "parallelism" means, but I would have expected it to mean the former, which means that a thread in control of creating commit trees would not count towards/be relevant to the `--jobs` flag.

Reply:
Ok I've adjusted it to a single user visible job. The implementation details may use a single or multiple threads, but that's open anyway.